bench/irmin-pack/tree.ml

```diff
@@ -90,247 +218,528 @@ module Parse_trace = struct
-      ~sampling_interval:50
+    let+ commit = Store.Commit.v repo ~info ~parents:parents_store tree in
+    Store.Tree.clear tree;
+    let h_store = Store.Commit.hash commit in
+    (* It's okey to have [h_trace] already in history. It corresponds to
+     * re-commiting the same thing, hence the [.replace] below. *)
+    Hashtbl.replace t.hash_corresps (unscope h_trace) h_store;
+    maybe_forget_hash t h_trace;
+    t.latest_commit <- Some h_store
+
+  let add_operations t repo operations n stats =
+    let rec aux l i =
+      match l with
+      | Bootstrap_trace.Checkout (h, ctx1) :: tl ->
+          exec_checkout t repo h ctx1 |> with_monitoring stats `Checkout
+          >>= fun () -> aux tl (i + 1)
+      | Add op :: tl ->
+          exec_add t op.key op.value op.ctx0 op.ctx1
+          |> with_monitoring stats `Add
+          >>= fun () -> aux tl (i + 1)
+      | Remove (keys, ctx0, ctx1) :: tl ->
+          exec_remove t keys ctx0 ctx1 |> with_monitoring stats `Remove
+          >>= fun () -> aux tl (i + 1)
+      | Copy op :: tl ->
+          exec_copy t op.key_src op.key_dst op.ctx0 op.ctx1
+          |> with_monitoring stats `Copy
+          >>= fun () -> aux tl (i + 1)
+      | Find (keys, b, ctx0) :: tl ->
+          exec_find t n i keys b ctx0 |> with_monitoring stats `Find
+          >>= fun () -> aux tl (i + 1)
+      | Mem (keys, b, ctx0) :: tl ->
+          exec_mem t n i keys b ctx0 |> with_monitoring stats `Mem >>= fun () ->
+          aux tl (i + 1)
+      | Mem_tree (keys, b, ctx0) :: tl ->
+          exec_mem_tree t n i keys b ctx0 |> with_monitoring stats `Mem_tree
+          >>= fun () -> aux tl (i + 1)
+      | [ Commit op ] ->
+          exec_commit t repo op.hash op.date op.message op.parents op.ctx0
+          |> with_monitoring stats `Commit
+      | Commit _ :: _ | [] ->
+          failwith "A batch of operation should end with a commit"
+    in
+    aux operations 0
+
+  let add_commits repo max_ncommits commit_seq on_commit on_end stats () =
+    with_progress_bar ~message:"Replaying trace" ~n:max_ncommits ~unit:"commits"
     @@ fun prog ->
-    let t = { tree = Store.Tree.empty } in
-    let rec array_iter_lwt prev_commit i =
-      if i >= n then Lwt.return_unit
-      else
-        let operations = commits.(i) in
-        let* _, prev_commit = add_operations t repo prev_commit operations i in
-        let* () = on_commit i (Option.get prev_commit) in
-        prog Int64.one;
-        array_iter_lwt prev_commit (i + 1)
+    let t =
+      {
+        contexts = Hashtbl.create 3;
+        hash_corresps = Hashtbl.create 3;
+        latest_commit = None;
+      }
     in
-    array_iter_lwt None 0
+
+    (* Manually add genesis context *)
+    Hashtbl.add t.contexts 0L { tree = Store.Tree.empty };
+
+    let rec aux commit_seq i =
+      match commit_seq () with
+      | Seq.Nil ->
+          (* Let's print the length of [data/commitments] using t.latest_commit
```

Comment:
why `data/commitments`? is it the widest?

Reply:
Yes!